### DISCUSSION TITLE
Displaying notifications while app is running in Android

### DISCUSSION THREAD

**Discussion by leo-petrucci** [ionic-team/capacitor#3660] [Q&A]
I managed to get Push notifications working as intended on Android, unfortunately these only work when the app is in the background or closed.

I found a couple of sources telling me that I need to handle these within the app like [this article](https://wajahatkarim.com/2018/05/firebase-notifications-in-background--foreground-in-android/) and [this answer](https://stackoverflow.com/questions/55446687/how-to-fix-notification-not-showing-up-when-app-is-running), but I have not been able to get these to work.

Where exactly should I be extending that class to handle notifications while the app is open? What packages do I need to import? Why doesn't the capacitor notification plugin handle these out of the box?

**Comment by leo-petrucci** [comment]
Still struggling with this, would really appreciate any help!

**Comment by AE1NS** [comment]
You could show an ion-alert when you are inside the app and receive a push notification, like:
```
PushNotifications.addListener('pushNotificationReceived', async (notification: PushNotification) => {
    await (
        await this._alertController.create({
            header: notification.title,
            message: notification.body,
            buttons: ['Close']
        })
    ).present();
});
```

**Reply by leo-petrucci** [reply]
Thank you for the reply! I tried this but I get a `Line 1 - Msg: Uncaught (in promise) TypeError: Cannot read property '_alertController' of undefined` error when receiving a notification.

Also wouldn't this just show a modal? Is there really no way to show a system notification inside my app without manually coding it myself?

**Reply by AE1NS** [reply]
You could use this guide to setup push notifications:
https://capacitorjs.com/docs/apis/push-notifications
https://capacitorjs.com/docs/guides/push-notifications-firebase

And then use my approach with the ionic alert. The variable _alertController is injected in the constructor.
```
constructor(private _alertController: AlertController) { }
```

**Reply by leo-petrucci** [reply]
I've got the Push notifications already setup, but I don't think I can get away with showing an alert rather than a notification :/

**Comment by pouriamoosavi** [comment]
I've managed to solve my problem by using another plugin called `local-notifications`: https://capacitorjs.com/docs/apis/local-notifications.

I call the `schedule` function from `pushNotificationReceived` listener on `PushNotifications` to shows a notification even when my app is foreground.
```typescript
import {
  PushNotificationSchema,
  PushNotifications,
} from '@capacitor/push-notifications';
import { LocalNotifications, ScheduleOptions } from '@capacitor/local-notifications';

// ...

PushNotifications.addListener(
  'pushNotificationReceived',
  async (notification: PushNotificationSchema) => {
    let not: ScheduleOptions = {
      notifications: [{
        id: Date.now(),
        body: notification.body,
        title: notification.title,
        ongoing: false,
      }]
    };
    const result = await LocalNotifications.schedule(not)
    console.log(result)
  }
);
 ```
Hope it helps

**Comment by pavelety** [comment]
I have the same problem only for Android, on iOS native push comes even with opened app, so I had to make a toast for Android:
```typescript
import {NavController, Platform, ToastController} from '@ionic/angular';
...
if (this.platform.is('android')) {
  PushNotifications.addListener('pushNotificationReceived', (notification: PushNotificationSchema) => {
      this.toastCtrl.create({
        message: `${notification.title}: ${notification.body}`,
        position: 'top',
        buttons: [{
          icon: 'chevron-forward-outline',
          side: 'start',
          role: 'show',
          handler: () => this.navCtrl.navigateForward(YOUR_URL)
        }, {
          icon: 'close',
          side: 'end',
          role: 'cancel'
        }]
      }).then(toast => toast.present());
    }
  );
}
```
